draft: treewide: dotnet reorg: dotnet-sdk -> dotnet

corngood opened the pull request:
## Description of changes

FYI @NixOS/dotnet 

This is an early draft of some potential package renames that @mdarocha and I discussed on a thread on #190129.

- There is no default version of dotnet (was 6 before).  I'm not sure if this is a good idea.  How many places that used e.g. `dotnet-sdk` would have been portable to other versions?

- There is no top-level `dotnet-runtime` / `dotnet-sdk`.  These may still be useful in some cases?

- Specifying aspnetcore as runtime is still awkward.  Maybe this should be a dedicated flag on `buildDotnetModule` or something like `buildAspnetcoreModule`? 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

corngood committed:
dotnet: reorg phase 1 - eval passes with aliases

corngood committed:
dotnet reorg phase 2 - aliases not required

mdarocha commented:
- I'd agree, we don't need a default `dotnetPackages` without the number 
- I'd still leave at least a top-level `dotnet-sdk`, since it may be useful for some ad-hoc usages (ie. with `nix run`)
- See https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging - I don't think it's awkward, and the current layout aligns with what's recommended by Microsoft

corngood commented:
> I'd still leave at least a top-level dotnet-sdk, since it may be useful for some ad-hoc usages (ie. with nix run)

Agreed.  I'll put some of those back in, but avoid using them in other packages.

> See https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging - I don't think it's awkward, and the current layout aligns with what's recommended by Microsoft

I'm not sure what you mean here, exactly.  Are you referencing what I said about aspnetcore being awkward?  By 'current' do you mean what's in nixpkgs, or in this PR?


I do have a couple of thoughts related to what's in that doc...

- Should we use e.g. `dotnet-sdk` instead of just `sdk`, even when in `dotnetPackages_[version]`?  It feels like repeating yourself, but it could be advantageous when using scopes.

- They use major and minor versions everywhere.  Should we do that instead of just major?  My instinct is to carry on with major only, until we have a reason to provide multiple minor versions.

mdarocha commented:
> > See https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging - I don't think it's awkward, and the current layout aligns with what's recommended by Microsoft
> 
> 
> 
> I'm not sure what you mean here, exactly.  Are you referencing what I said about aspnetcore being awkward?  By 'current' do you mean what's in nixpkgs, or in this PR?
> 

I was referring to your comment about using aspnetcore runtime being awkward - the current setup we have is pretty alright I think, and we should leave it as is.

> I do have a couple of thoughts related to what's in that doc...
> 
> 
> 
> - Should we use e.g. `dotnet-sdk` instead of just `sdk`, even when in `dotnetPackages_[version]`?  It feels like repeating yourself, but it could be advantageous when using scopes.
> 
I agree, it feels like repeating yourself. I don't think better naming in scopes is enough of an advantage.

> 
> - They use major and minor versions everywhere.  Should we do that instead of just major?  My instinct is to carry on with major only, until we have a reason to provide multiple minor versions.

I'd stick with major only, and add a minor version when there'll be a need for it.

raphaelr commented on pkgs/top-level/all-packages.nix:
```diff
@@ -915,30 +915,22 @@ with pkgs;
-
-  buildDotnetModule = callPackage ../build-support/dotnet/build-dotnet-module { };
-  nuget-to-nix = callPackage ../build-support/dotnet/nuget-to-nix { };
+  buildDotnetModule = callPackage ../build-support/dotnet/build-dotnet-module { dotnetPackages = dotnetPackages_6; };
+  nuget-to-nix = callPackage ../build-support/dotnet/nuget-to-nix { dotnetPackages = dotnetPackages_6; };
```
One question, I don't have strong feelings about this in either direction: What is the use case for having a top-level `buildDotnetModule`, `nuget-to-nix`, and `buildDotnetGlobalTool`?

corngood replied:
I don't think there's a reason to keep them.  I can move them to aliases.

GGG-KILLER replied:
I believe it'd be better to just outright move then to aliases as a warning.

raphaelr reviewed (commented):
> Specifying aspnetcore as runtime is still awkward. Maybe this should be a dedicated flag on buildDotnetModule or something like buildAspnetcoreModule?

I'd go with a dedicated flag:
- Default runtime is the normal runtime
- `buildDotnetModule { runtime = "aspnetcore"; }` selects the aspnetcore runtime
- `buildDotnetModule { runtime = "sdk"; }` uses the SDK as the runtime (used by fsautocomplete, but maybe we can have it use the normal runtime instead)

corngood commented:
FYI @GGG-KILLER.  I just realised you aren't in the dotnet maintainers team.  If you'd like to be pinged on dotnet architectural stuff, you could make a PR adding yourself to the team.

GGG-KILLER commented:
> FYI @GGG-KILLER.  I just realised you aren't in the dotnet maintainers team.  If you'd like to be pinged on dotnet architectural stuff, you could make a PR adding yourself to the team.

Oh thanks, will look into it! If you have any guides or examples on how to do it I'd be very grateful!

I hadn't added myself before because there were more active maintainers and I didn't feel like I was worthy enough to be on the team but I think I've matured enough to be part of it now.

corngood commented:
a7dfee8bdd304abf60b3ccc98265457a4be07078 is a random example of someone adding themselves to the maintainers list.

e3ca8030852cd391fdba9eed9dd3da88c83e9538 is one for adding someone to a team.

I think you could just make a PR with two commits like those.  There's some stuff in maintainers/README.md, but it all looks pretty straightforward.

GGG-KILLER commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -30,16 +29,24 @@ let
+  self = {
+    inherit systemToDotnetRid;
+
+    combinePackages = attrs: callPackage (import ./combine-packages.nix attrs) {};
```
I believe something we can do here is `withExtraSdks` and then imply the dotnet that this is part of as the CLI (first arg).

e.g.: `dotnetPackages_9.withExtraSdks [dotnetPackages_8.sdk]` would be the same as `dotnetPackages_9.combinePackages [dotnetPackages_9.sdk dotnetPackages_8.sdk]`.

corngood replied:
Done in latest.

GGG-KILLER commented on pkgs/servers/bililiverecorder/default.nix:
```diff
@@ -2,15 +2,15 @@
-  dotnet = with dotnetCorePackages; combinePackages [
-    runtime_6_0
-    aspnetcore_6_0
+  dotnet = with dotnetPackages_6; combinePackages [
+    runtime
+    aspnetcore
```
I know this is how the package was initially conceived, but aspnetcore already implies runtime, it's just runtime with the extra things needed for ASP.NET Core, no?

corngood replied:
Yeah, that's my understanding.

corngood commented:
I think I'd like to take a different approach to this and not do top level attributes by version (e.g. `dotnetPackages_8`).  That would mean either using `dotnetCorePackages` or `dotnetPackages`.  I think long-term it would be nice to combine `dotnetCorePackages` into `dotnetPackages` regardless.

I'd still like to have one attribute that represents the "dotnet distribution".  I'm not sure exactly what to call this, but I mean the SDK and corresponding runtime, aspnetcore and all the various other packages that are covered in https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging.  I'm thinking `dotnet_[version][-bin].[sdk|runtime|aspnetcore]`.  This is how the source-built packages are currently named.

Combining multiple sdk/runtime versions into one tree is something that's supported upstream, so I think it makes sense to have support for that outside of the versioned attributes. (e.g. `dotnetCorePackages.combinePackages`).

When I get a chance I'll update this PR to match what I'm proposing.

corngood commented:
I did another pass on this.  It evaluates, but some things probably don't build correctly.

- I added top-level attributes for `dotnet_6/7/8/9[-bin]`.

- I added `withExtraSDKs` as suggested by @GGG-KILLER.  This covers most of the existing uses of `combinePackages`.

I think this looks pretty clean from the POV of a package using it.  For example:

```
{ dotnet_8, dotnet_6 }:
let
  dotnet = dotnet_8.withExtraSDKs [ dotnet_6.sdk ];

in dotnet.buildDotnetModule rec {
```

GGG-KILLER commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -30,21 +27,57 @@ let
+
+    inherit combinePackages;
+
+    ## Files in versions/ are generated automatically by update.sh ##
+    dotnet_6-bin = makeDotnet (import ./versions/6.0.nix buildAttrs);
+    dotnet_7-bin = makeDotnet (import ./versions/7.0.nix buildAttrs);
+    dotnet_8-bin = makeDotnet (import ./versions/8.0.nix buildAttrs);
+    dotnet_8_0_102-bin = makeDotnet (import ./versions/8.0.102.nix buildAttrs);
+    dotnet_9-bin = makeDotnet (import ./versions/9.0.nix buildAttrs);
+
+    # We don't have source-builds of 6/7
+    dotnet_6 = dotnet_6-bin;
+    dotnet_7 = dotnet_7-bin;
+    dotnet_8 = makeDotnet (callPackage ./8 { bootstrapSdk = dotnet_8_0_102-bin.sdk; });
+    dotnet_9 = makeDotnet (callPackage ./9 {});
```
I'm a bit on the fence about this because of .NET 8, we'd be providing an intuitive default (`dotnet_8`) which is actually just a maintenance version instead of a more featureful one.

Maybe for the time being, we suffix the source build ones and not the binary ones? Or prefix both

corngood replied:
I think we should probably discourage use of the binary packages, especially due to #169427.

GGG-KILLER replied:
I mean, according to what omajid said:

> The LICENSE file for the linux SDK tarballs, as far as I am aware, is not the proprietary EULA.

We don't need to worry about that since we only handle mac and linux, if we handled windows then that'd be a concern.

I just downloaded every runtime, SDK and ASP.NET Core runtime tarball from MS for the latest versions of .NET 6, 7 and 8 and am reading through the licenses and I don't see any proprietary ones so far.

corngood replied:
Ah, okay, maybe we should close that then.

GGG-KILLER replied:
I've left a comment there noting this out, I'll leave it up to the people already in there to decide whether to keep it open or not, but thanks for bringing it up

corngood replied:
I still think it would be a good idea to use source-built packages as much as possible.  It should allow us to eventually make the dotnet ecosystem more nix-friendly.  For example, dropping the elf patching stuff for system libs.  We could also add support for locating sdks, runtimes, and nuget packages from multiple locations in /nix/store.

I do share your concern about the out of date release though.  There's also a question over cross-compilation, because the source built SDKs only have packages for the target platform, and I haven't been able to build them in cross yet.

GGG-KILLER replied:
> I still think it would be a good idea to use source-built packages as much as possible. It should allow us to eventually make the dotnet ecosystem more nix-friendly.

Yes, I am in favor of that for .NET 9, however for .NET 8 I think it'd be better for the user to have a more featureful SDK than one that is source-built when they have no visible benefit (since we already do patching for the binary one, I don't think they'll see much difference).

omajid replied:
In case it's not obvious (it wasn't to me at first), please note that 8.0.1xx and 8.0.2xx are both maintained in parallel. Both are getting bugfixes.

In fact, 8.0.1xx will be maintained for the entire duration of .NET 8's lifecycle, while 8.0.2xx will be EOL'ed by upstream when the associated Visual Studio version goes EOL. There would be a 8.0.3xx (and possibly 8.0.4xx, and so on) to replace it, though.

GGG-KILLER replied:
I believe that the packages without a minor version (e.g. `dotnet_8`) should point to the latest SDK (e.g. 8.0.2xx or 8.0.4xx, whichever is latest).

I'm aware that 8.0.1xx will be maintained throughout the lifecycle but I think users will prefer to have the latest version as opposed to just the initial one but with bugfixes.

corngood replied:
> I think users will prefer to have the latest version as opposed to just the initial one but with bugfixes.

We could make the latest one usable via e.g. `dotnet-sdk_8`, but still use the source-built one inside nixpkgs wherever possible.

There's lots of precedent for `-bin` packages, but nothing like `-src` that I can see.  I'm not sure if there's anything where the latest version is only available in `-bin`.

mdarocha commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -30,21 +27,57 @@ let
+    dotnet_6-bin = makeDotnet (import ./versions/6.0.nix buildAttrs);
+    dotnet_7-bin = makeDotnet (import ./versions/7.0.nix buildAttrs);
+    dotnet_8-bin = makeDotnet (import ./versions/8.0.nix buildAttrs);
+    dotnet_8_0_102-bin = makeDotnet (import ./versions/8.0.102.nix buildAttrs);
```
maybe name this `dotnet_8_bootstrap` or something similar, so the purpose of the package is obvious (and it doesn't get used when not needed by accident). Or at least add a comment.

js6pak replied:
I don't see why they would be different versions at all. .NET 8 SDK 8.0.201 and 8.0.102 are both part of the .NET 8.0.2 release. I think the best option would be exposing all feature bands as `dotnet_8-bin.sdk_x` (so currently `dotnet_8-bin.sdk_1` and `dotnet_8-bin.sdk_2`) and alias `dotnet_8-bin.sdk` to whatever the latest feature band currently is. We can then drop feature bands [as they become EOL](https://learn.microsoft.com/en-us/dotnet/core/porting/versioning-sdk-msbuild-vs#lifecycle) by throwing like we do with <=5.0 sdks right now.

This has the added benefit for people using nixpkgs to manage their development enviroments as .NET SDK feature bands can have major new features and bugs, see https://github.com/dotnet/sdk/issues/39142 which currently breaks most of the style analyzers in the `2xx` feature band while it still works in the `1xx`.

corngood replied:
This is a pretty good idea.  8.0.3 just came out, so I'll see if i can teach the update script to pull down each SDK for a corresponding runtime.

corngood replied:
I made a PR with the feature band idea:  #299822

corngood commented:
> @corngood With the feature band PR, will we be moving the source build to the band it supports and leave the generic to the latest (binary in .NET 8's case) one?

I'd like to, but there's still at least one problem.  The source-built sdk only has packages for the target rid, and we currently do fetch-deps by restoring for all possible rids, so you end up with a bunch of stuff in deps.nix that wouldn't be there with the binary sdk.  I should probably make an issue to discuss this and possible solutions.

GGG-KILLER commented:
> I'd like to, but there's still at least one problem. The source-built sdk only has packages for the target rid, and we currently do fetch-deps by restoring for all possible rids, so you end up with a bunch of stuff in deps.nix that wouldn't be there with the binary sdk. I should probably make an issue to discuss this and possible solutions.

Given that the source-built distribution *should* be able to cross-build, I believe we should find a way to add to it the packages for other RIDs.
Is it a matter of building it for 4 different RIDs or are these the bootstrap pre-built CentOS packages?

corngood commented:
> Is it a matter of building it for 4 different RIDs or are these the bootstrap pre-built CentOS packages?

The bootstrap packages are also only for one RID, I believe, but it's the `.packages` ones that are the problem.  For source sdks they are built with the sdk, and for binary sdks they are pulled from nuget using the list in `update.sh`.

> Given that the source-built distribution should be able to cross-build, I believe we should find a way to add to it the packages for other RIDs.

I think there are two main routes we could go down for `fetch-deps`:

1. restore for all RIDs and track dependencies for all of them
2. restore only the current RID and infer the dependencies for the others

(1) is what we currently do for the default `fetch-deps`, and (2) is what we do for the source-built sdk `fetch-deps`.

Assuming we continue with (1), we need a way to make `fetch-deps` aware of the available packages on all RIDs.  There are a couple ways we could do this:

- cross-build and use those (I did a little testing of cross-built sdks, and it was not very promising)
- use the binary (nuget) ones

Some other random thoughts:

- we should probably be filtering dependencies by RID to reduce the build closure size
- what happens if you depend on another (nix) package for dependencies, but it can't be built in cross?

The best option for short-term progress is probably pulling in the nuget packages for the cross targets, but only in `fetch-deps`, not actual builds.

GGG-KILLER commented:
> The bootstrap packages are also only for one RID, I believe, but it's the `.packages` ones that are the problem.  For source sdks they are built with the sdk, and for binary sdks they are pulled from nuget using the list in `update.sh`.

I see, and in this case would we be able to pull from nugget for the other RIDs? Given that we do not have a way of joining build results from different builders in Nix and even if we had, that would totally rule out the people who want to manually build it from scratch.

> I think there are two main routes we could go down for `fetch-deps`:
> 
> 1. restore for all RIDs and track dependencies for all of them
> 2. restore only the current RID and infer the dependencies for the others
> 
> (1) is what we currently do for the default `fetch-deps`, and (2) is what we do for the source-built sdk `fetch-deps`.
> 
> Assuming we continue with (1), we need a way to make `fetch-deps` aware of the available packages on all RIDs.  There are a couple ways we could do this:l
> 
> - cross-build and use those (I did a little testing of cross-built sdks, and it was not very promising)
> - use the binary (nuget) ones
> 
> Some other random thoughts:
> 
> - we should probably be filtering dependencies by RID to reduce the build closure size
> - what happens if you depend on another (nix) package for dependencies, but it can't be built in cross?
> 
> The best option for short-term progress is probably pulling in the nuget packages for the cross targets, but only in `fetch-deps`, not actual builds.
>  

Here I believe we have a few issues as well:

1. If we filter the deps by RID, wouldn't that mean that we'd have a different `deps.nix` per RID?
  How would we detect for which RID a given package is?
  Wouldn't that require a modification or even a rewrite of `nuget-to-nix`?
2. Is `fetch-deps` really the only thing we have to worry about? What about command line cross builds?

Other than these I believe restoring just the packages for the target system would indeed be ideal and reduce closure size significantly.

GGG-KILLER commented:
> Is it worth the trouble? I don't know. Maybe there's a fundamentally better way of doing all this that doesn't involve looking at what got downloaded during a restore.

I suggested this before, but I think we could use the NuGet packages to get information from the packages that are in the project and manage them in a smarter way.
However this *would* involve having to rewrite nuget-to-nix.
 
> Do we want all SDKs to support cross targeting to all RIDs like the binary ones do?I

Yes, at least I develop windows tools in Linux and I believe other people would want to be able to develop for all RIDs.
If it's something we offer in the binary releases, we should have feature parity with it for the source build in my opinion.

corngood commented:
> I suggested this before, but I think we could use the NuGet packages to get information from the packages that are in the project and manage them in a smarter way.

Yeah, that would be great if possible.  However I've seen some pretty crazy shit in msbuild projects (mostly sdk/runtime related ones to be fair) where nuget is called from custom targets in order to restore specific packages, etc.

> If it's something we offer in the binary releases, we should have feature parity with it for the source build in my opinion.

So that means we need to build `Microsoft.NETCore.App.Runtime.win-x64` et al from source on all nix platforms.  I'll give the cross build another try at some point.

GGG-KILLER commented:
> Yeah, that would be great if possible. However I've seen some pretty crazy shit in msbuild projects (mostly sdk/runtime related ones to be fair) where nuget is called from custom targets in order to restore specific packages, etc.

In that extreme case, then people would have to manually list the deps with us providing some kind of fallback for it (maybe load from a directory or something).
It'd indeed be something we have to think of. Maybe initially doing a restore to different directories for each RID would be the best option.

> So that means we need to build `Microsoft.NETCore.App.Runtime.win-x64` et al from source on all nix platforms. I'll give the cross build another try at some point.

Yeah, it'd be a rather big build but we could maybe provide an input flag to the derivation so that people can `override` it to not get such a big closure size (and for internal nixpkgs usage).
